activesupport/lib/active_support/cache/memory_store.rb

```diff
@@ -114,45 +131,44 @@ def synchronize(&block) # :nodoc:
       private
         PER_ENTRY_OVERHEAD = 240
 
-        def cached_size(key, entry)
-          key.to_s.bytesize + entry.size + PER_ENTRY_OVERHEAD
+        def cached_size(key, payload)
+          key.to_s.bytesize + payload.bytesize + PER_ENTRY_OVERHEAD
         end
 
         def read_entry(key, **options)
           entry = nil
           synchronize do
-            entry = @data.delete(key)
-            if entry
-              @data[key] = entry
+            payload = @data.delete(key)
+            if payload
+              @data[key] = payload
+              entry = deserialize_entry(payload)
               entry = entry.dup
             end
           end
-          entry&.dup_value!
           entry
         end
 
         def write_entry(key, entry, **options)
-          entry.dup_value!
+          payload = serialize_entry(entry)
           synchronize do
-            return false if options[:unless_exist] && @data.key?(key)
-
-            old_entry = @data.delete(key)
-            if old_entry
-              @cache_size -= (old_entry.size - entry.size)
+            return false if @data.key?(key) && options[:unless_exist]
```

Comment:
There was a blank line after the guard clause here before, let's add it back in.

```suggestion
            return false if @data.key?(key) && options[:unless_exist]

```

Any reason to swap the order of the checks here?

Reply:
> Any reason to swap the order of the checks here?

No, actually I should switch them back.